README.md
# walter

Walter é um programa escrito em Python feito para manter a comunicação entre dois computadores, dependendo somente de uma rede local com a porta 39 TCP.

Para iniciar um servidor:
```
 python server.py [ip] [senha]
```

Para iniciar um cliente:
```
 python client.py
```

Até agora o Walter não necessita de alguma dependência especial, somente das bibliotecas padrão. ~~espero que você tenha se não te mato.~~


walter/changelog.txt
0.0.1 to 0.2 [OBRIGATORY]
	DIFFIE-(HELL)[MANS] IN WALTER(the secret is now the encryp. password)!
		-> because of that encryption passwors are now useless
		-> END-TO-END encryption
			-Walter server generates a new password for each new client
			-not using a defined password by the server operator
		-> SHA256 is used instead of MD5 to hash passwords

walter/wcr2.py
#internal libraries
import random
import os
import base64
import math

from collections import Counter
BLOCKSIZ = 16

def xor_strings(s,t):
    return "".join(chr(ord(a)^ord(b)) for a,b in zip(s,t))

def entropy(s):
    p, lns = Counter(s), float(len(s))
    return -sum( count/lns * math.log(count/lns, 2) for count in p.values())

'''
    q1w2e3r4
    q1w2e3r4q1w2e3r4q1w2e3r4q1w2e3r[4]
'''
def pad(passwd, l):
    plen = l - (len(passwd) % l)
    padding = chr(0) * plen
    return passwd + padding

class BException(Exception):
    def __init__(self, value):
        self.value = value
    def __str__(self):
        return repr(self.value)

class LengthError(BException):
    def __init__(self, value):
        BException.__init__(self, value)

class EntropyError(BException):
    def __init__(self, value):
        BException.__init__(self, value)

#WCR principal class
class WCR:
    def __init__(self, length):
        if length >= 1024 and length % 256 == 0:
            self.key = os.urandom(length)
        else:
            raise LengthError("length must be a multiple of 256 and >= 1024")
        self.length = len(self.key)
    def encrypt(self, message, passwd):
        '''Encrypt a piece of data with WCR'''
        if entropy(passwd) <= 2.2:
            raise EntropyError("password's entropy must be > 2.2")
        text = list(message)
        # q = random.SystemRandom().randint(0, len(passwd)-1)
        # k = passwd[q]
        i = 0
        pwdpadded = pad(passwd, self.length)
        for c in text:
            # n = ord(self.key[i]) * ord(k)
            # new_character = chr(n % 255)
            newchar = xor_strings(self.key[i], pwdpadded[i])
            text[i] = xor_strings(text[i], newchar)
            i += 1
        return ''.join(text)
    def decrypt(self, ciphertext, passwd):
        '''Decrypt a piece of data with WCR'''
        text = list(ciphertext)
        #k = passwd[q]
        i = 0
        pwdpadded = pad(passwd, self.length)
        for c in text:
            # n = ord(self.key[i]) * ord(k)
            # new_character = chr(n % 255)
            newchar = xor_strings(self.key[i], pwdpadded[i])
            text[i] = xor_strings(text[i], newchar)
            i += 1
        return ''.join(text)
    def export(self):
        '''export the key in base64 format'''
        return base64.b64encode(self.key)
    def import_base64(self, new_iv):
        '''import a key in base64 format'''
        self.key = base64.b64decode(new_iv)
        self.length = len(self.key)


walter/error.py
# -*- coing: utf-8 -*-
# error.py -- walter error codes

import sys

err_codes = {
    # 0xx -- connection errors
    # 1xx -- client errors
    # 2xx -- server errors
    # 3xx -- diffie hellman key exchange errors
    # 4xx -- internal errors
    # 5xx -- crypto errors
    # 999 -- special error
    'SocketError': 001,
    'DH_NValid_PA': 301,
    'DH_NValid_PB': 302,
    'DH_NValid_DIFFIE': 303,
    'DH_NValid_YD': 304,
    
}

def err(err_code, comment=False):
    if not comment:
        comment = ''
    if isinstance(err_code, str):
        try:
            code = err_codes[err_code]
        except KeyError:
            code = 999
    else:
        code = err_code
    sys.stderr.write('Walter Error!\n')
    sys.stderr.write('Error Code: %d\n' % code)
    sys.stderr.write('Error Name: %s\n' % err_code)
    sys.stderr.write('Error COMM: %s\n' % comment)


walter/client.py
# -*- coding: utf-8 -*-
# client.py - Walter Client
import socket
import time
import thread
import sys
import hashlib
import ast

import diffie as dh
import error

from wcr2 import WCR

VERSION = '0.2'
AUTHOR = 'Lukas Mendes'
PORT = 39
BUFSIZE = 4096

def passwordHash(x):
    return hashlib.sha256(x).hexdigest()

def printout(msg):
    sys.stdout.write(msg) ; sys.stdout.flush()

def socksend(sock, message):
    try:
        sock.send(message)
    except socket.error as e:
        error.err('SocketError', e.message)

def receber_msg(n, sock, crypt_obj, ep):
    while True:
        recieved = sock.recv(BUFSIZE)
        #r = ast.literal_eval(recieved)
        decrypted = crypt_obj.decrypt(recieved, ep)
        printout('\rmsg: %s\n' % decrypted)
        printout('%s:msg> ' % n)
    thread.exit()

def main():
    print "Walter Client v%s by %s" % (VERSION, AUTHOR)
    
    tcp = socket.socket()
    addr = raw_input('IP do Servidor: ')
    CONN = (addr, PORT)
    trypwd = raw_input('Senha do Servidor: ')
    nick = raw_input('Nick do Usuário: ')
    
    tcp.connect(CONN)
    socksend(tcp, passwordHash(trypwd))
    # handling diffie-hellman
    dh_recep = tcp.recv(BUFSIZE)
    
    # handling wrong passwords
    if dh_recep == "senha errada, seu merda.":
        print 'Wrong password!'
        print 'closing...'
        tcp.close()
        return 1
    
    if dh_recep.startswith('?DIFFIE'):
        D = dh_recep.split()
        prime = int(D[1])
        generator = int(D[2])
        socksend(tcp, '?YESDIFFIE')
        a = dh.get_rand(8)
        PA = dh.g(generator, a, prime)
        socksend(tcp, '?PUBLICA %d' % PA)
        _PB = tcp.recv(BUFSIZE)
        if not _PB.startswith('?PUBLICB'):
            error.err('DH_NValid_PB')
            return 1
        PB = long(_PB.split()[1])
        S = dh.g(PB, a, prime)
        epwd = S
    else:
        error.err('DH_NValid_DIFFIE')
        return 1
    
    # getting WCR key
    H = tcp.recv(BUFSIZE)
    h = H.split(';')
    print h[0]
    key = h[1]
    
    w = WCR(2048)
    w.import_base64(key)
    epwd = str(epwd)
    thread.start_new_thread(receber_msg, tuple([nick, tcp, w, epwd]))
    
    print 'Bem-vindo a um servidor do Walter, boas conversas!'
    
    while True:
        msg = raw_input()
        if msg == '/exit': # closing connection
            socksend(tcp, '1CLOSE')
            tcp.close()
            return 0
        r = w.encrypt('%s: %s' % (nick, msg), epwd)
        socksend(tcp, str(r))

if __name__ == '__main__':
    sys.exit(main())


walter/diffie.py
#python
# diffie.py
# implements Diffie-Hellman Key Exchange

import random
import os
import time
import sys
sys.setrecursionlimit(7000)

PRIME = 127

def get_rand(n):
    return int(os.urandom(n).encode('hex'), 16)

def g(gen, x, prime):
    return pow(gen, x, prime)

def setBit(int_type, offset):
    mask = 1 << offset
    return(int_type | mask)

_mrpt_num_trials = 5 # number of bases to test

def test_prime(n):
    assert n >= 2
    # special case 2
    if n == 2:
        return True
    # ensure n is odd
    if n % 2 == 0:
        return False
    # write n-1 as 2**s * d
    # repeatedly try to divide n-1 by 2
    s = 0
    d = n-1
    while True:
        quotient, remainder = divmod(d, 2)
        if remainder == 1:
            break
        s += 1
        d = quotient
    assert(2**s * d == n-1)
 
    # test the base a to see whether it is a witness for the compositeness of n
    def try_composite(a):
        if pow(a, d, n) == 1:
            return False
        for i in range(s):
            if pow(a, 2**i * d, n) == n-1:
                return False
        return True # n is definitely composite
 
    for i in range(_mrpt_num_trials):
        a = random.randrange(2, n)
        if try_composite(a):
            return False
 
    return True # no base tested showed n as composite

def get_tprime(n):
    x = get_rand(n)
    setBit(x, 0)
    setBit(x, long(x).bit_length())
    return x

def test_tprime(x):
    b = long(x).bit_length()
    c1 = test_prime(x)
    c2 = pow(2, b-1) < x < pow(2, b)
    return c1 and c2

def get_prime(x):
    p = get_tprime(x)
    while not test_tprime(p):
        p = get_tprime(x)
    return p

def test():
    p = get_tprime(32)
    while not test_tprime(p):
        p = get_tprime(32)
    gn = get_rand(64) # base generator
    a = get_rand(8)
    b = get_rand(8)
    print p, gn, a, b
    PA = g(gn, a, p)
    PB = g(gn, b, p)
    print 'PA =', PA
    print 'PB =', PB
    SA = g(PB, a, p)
    SB = g(PA, b, p)
    print 'SA =', SA
    print 'SB =', SB
    print 'SA == SB ->', SA == SB

walter/server.py
# -*- coding: utf-8 -*-
# server.py - Walter Server
import socket
import time
import thread
import sys
import hashlib
import ast

import diffie as dh
import error

from wcr2 import WCR

# coisa q tenho q faze
# log no servidor

VERSION = '0.2'
AUTHOR = 'Lukas Mendes'
PORT = 39
BUFSIZE = 4096

sockets = []
secrets = {}
nicks = {}

def passwordHash(x):
    return hashlib.sha256(x).hexdigest()

def printout(msg):
    sys.stdout.write(msg); sys.stdout.flush()

def socksend(sock, message):
    try:
        sock.send(message)
    except socket.error as e:
        error.err('SocketError', e.message)

def broadcast_all(start_socket, message):
    for s in sockets:
        if s == start_socket:
            pass
        else:
            try:
                printout('\n broadcasting: %s\n' % repr(s))
                socksend(s, message)
            except:
                printout("\n error: %s\n" % repr(s))
                try: s.close()
                except: pass
                sockets.remove(s)

def broadcast_encryp(sock, m, crypt_obj):
    global secrets
    global sockets
    for s in sockets:
        if s == sock:
            continue
        ssecret = secrets[hash(s)]
        new_message = crypt_obj.encrypt(m, ssecret)
        socksend(s, str(new_message))

def handle_diffie(sock, client):
    '''Handles DHKE(Diffie-Hellman Key Exchange)'''
    gen = dh.get_rand(64)
    p = dh.get_prime(32)
    socksend(sock, '?DIFFIE %d %d' % (p, gen))
    resp = sock.recv(BUFSIZE)
    if resp == '?YESDIFFIE':
        b = dh.get_rand(8)
        PB = dh.g(gen, b, p)
        socksend(sock, '?PUBLICB %d' % PB)
        _PA = sock.recv(BUFSIZE)
        if not _PA.startswith('?PUBLICA'):
            error.err('DH_NValid_PA', client)
            return False
        PA = long(_PA.split()[1])
        S = dh.g(PA, b, p)
        return S
    else:
        error.err('DH_NValid_YD', client)
        return False

def conectado(sock, cli, spwd, skey, crypt_obj):
    global secrets
    global sockets
    global nicks
    cli = str(cli)
    print 'thread iniciado: %s' % cli
    sockets.append(sock)
    trypass = sock.recv(BUFSIZE)
    if trypass == passwordHash(spwd): #acertou
        print '@@@accept %s' % cli
        secret = handle_diffie(sock, cli[0])
        if not secret:
            print '##ERROR NOT SECRET %s' % cli
            sock.close()
            thread.exit()
        socksend(sock, "senha aceita;%s" % skey)
        secret = str(secret)
        secrets[hash(sock)] = secret
        while True:
            recieved_message = sock.recv(BUFSIZE)
            if recieved_message == '1CLOSE':
                cls_msg = '**SERVER**: %s closed connection' % nicks[hash(sock)]
                broadcast_encryp(sock, cls_msg, crypt_obj)
                print 'closing: %s' % repr(sock)
                break
            # e = ast.literal_eval(recieved_message)
            # plz dont do this
            # e = eval(recieved_message, {'secrets': secrets})
            m = crypt_obj.decrypt(recieved_message, secret)
            print 'recv from %s: %s' % (cli, repr(recieved_message))
            print '\tMESSAGE: %s' % m
            nicks[hash(sock)] = m.split(':')[0]
            broadcast_encryp(sock, m, crypt_obj)
        print '!!!fechado!!:', cli
        sockets.remove(sock)
        del secrets[hash(sock)]
        del nicks[hash(sock)]
        sock.close()
    else:
        sockets.remove(sock)
        print '###wrong_password: %s' % cli
        socksend(sock, "senha errada, seu merda.") # xinga fortemente o cara que errou a senha
        sock.close()
    print '###closing_thread: %s' % cli
    del sock
    thread.exit()

def main():
    print "Walter Server v%s by %s" % (VERSION, AUTHOR)
    argv = sys.argv
    default = False
    if len(argv) == 1 or argv[1] == 'deflt':
        default = True
        print 'running default arguments'
        print 'HOST -> localhost'
        print 'password -> 123'
        HOST, s_passwd = 'localhost', '123'
    if len(argv) != 4 and not default:
        print 'usage: [sudo] python %s HOST s_passwd' % argv[0]
        return 1
    if not default:
        HOST = argv[1]
        s_passwd = argv[2]
    w = WCR(2048)
    s_key = w.export()
    tcp = socket.socket()
    tcp.bind((HOST, PORT))
    tcp.listen(15)
    print 'Waiting for clients in %s...' % HOST
    while True:
        con, cliente = tcp.accept()
        thread.start_new_thread(conectado, tuple([con, cliente, s_passwd, s_key, w]))

if __name__ == '__main__':
    sys.exit(main())
